python/triton/runtime/jit.py

```diff
@@ -300,130 +375,157 @@ def _conclude_device_type(self, device_types: List[str], pinned_memory_flags: Li
-            grid_2 = grid[2] if grid_size > 2 else 1
-            if device_type is None:
-                device_types = [self._device_of(arg) for arg in regular_args_v(args_proxy)]
-                device_types = [_device_type for _device_type in device_types if _device_type != '']
-                device_type = self._conclude_device_type(device_types, [self._pinned_memory_of(arg) for arg in
-                                                                        regular_args_v(args_proxy)])
-
-            device_backend = None
-            if device_type not in ['cuda']:
-                device_backend = get_backend(device_type)
-                if device_backend is None:
-                    raise ValueError('Cannot find backend for ' + device_type)
-
-            if device is None:
-                if device_type in ['cuda']:
-                    device = get_current_device()
-                    set_current_device(device)
-                else:
-                    device = device_backend.get_current_device()
-                    device_backend.set_current_device(device)
-            if stream is None and not warmup:
-                if device_type in ['cuda']:
-                    stream = get_cuda_stream(device)
-                else:
-                    stream = device_backend.get_stream()
-
-            if num_warps is None:
-                num_warps = get_arch_default_num_warps(device_type)
-            if num_stages is None:
-                num_stages = get_arch_default_num_stages(device_type)
-
+    def run(self, *args, **kwargs):
+        from ..compiler import (CompiledKernel, compile, get_arch_default_num_stages,
+                                get_arch_default_num_warps)
+
+        # Get a compiler-flags arg like `num_warps` and remove it from kwargs.
+        def get_special_arg(name: str, default=None):
+            if name not in kwargs:
+                return default
+            ret = kwargs[name]
+            del kwargs[name]
+            return ret
+
+        grid = get_special_arg('grid')
+        num_warps = get_special_arg('num_warps')
+        num_ctas = get_special_arg('num_ctas', 1)
+        num_stages = get_special_arg('num_stages')
+        enable_warp_specialization = get_special_arg('enable_warp_specialization', False)
+        enable_fp_fusion = get_special_arg('enable_fp_fusion', True)
+        extern_libs = get_special_arg('extern_libs')
+        stream = get_special_arg('stream')
+        warmup = get_special_arg('warmup', False)
+        device = get_special_arg('device')
+        device_type = get_special_arg('device_type')
+
+        # Bind the remaining arguments to `fn`.
+        bound_args = self.signature.bind(*args, **kwargs)
+        bound_args.apply_defaults()
+
+        params_dict = {param.name: param for param in self.params}
+        args = [
+            KernelArg(name, value, params_dict.get(name, None))
+            for name, value in bound_args.arguments.items()
+        ]
```

Comment:
I do worry that the new logic will have a non-trivial amount of host overhead on each call